185011051 closed client access

ZachCMP opened the pull request:
https://www.pivotaltracker.com/story/show/184986956

Relies on frontend PR: https://github.com/greenriver/hmis-frontend/pull/238

ZachCMP committed:
Client now uses ACLs for permissions

ZachCMP committed:
WIP

ZachCMP committed:
Fixup permissions to spec

ZachCMP committed:
Merge branch 'release-73' into 185011051-closed-client-access

gigxz committed:
Add projects_including_wip fn

gigxz commented on drivers/hmis/app/models/hmis/user.rb:
```diff
@@ -85,12 +90,12 @@ def lock_access!(opts = {})
 
   private def permissions_base_for_entity(entity)
-    return [entity.data_source, *entity.projects] if entity.is_a? Hmis::Hud::Client
+    return entity.projects if entity.is_a? Hmis::Hud::Client
```
I think there is a bug here that was overlooked before. We should be taking into account projects at WIP enrollments here. I propose a `client.projects_including_wip` function https://github.com/greenriver/hmis-warehouse/pull/3004 that could be used here

gigxz commented on drivers/hmis/app/models/hmis/user.rb:
```diff
@@ -60,12 +60,17 @@ def load_effective_permissions
     define_method("#{permission}_for?") do |entity|
       return false unless send("#{permission}?")
 
+      # Use global permission scope if client is unenrolled
+      return send("#{permission}?") if entity.is_a?(Hmis::Hud::Client) && entity.projects.empty?
```
@eanders what do you think about whether to take WIP Enrollments into consideration when determining whether a client is "unenrolled" or not? I think if that enrollment has been started, the client access should probably be limited right away. 

@ZachCMP I didn't think through allll the implications when you asked in PT, sorry this parts a bit tricky. Clients that only have WIP enrollments are considered "unenrolled" for _reporting_ purposes, but for access purposes I think we should consider them enrolled. See https://github.com/greenriver/hmis-warehouse/pull/3004 for how this could be updated to use `projects_including_wip` if so. Lets wait to hear what Elliot thinks though.

eanders replied:
Agreed. WIP should be treated as an enrollment for access purposes.

ZachCMP replied:
Okay, so we **should** count clients as enrolled as soon as they have an enrollment, even if it's incomplete and the intake assessment is in progress, correct?

gigxz replied:
Yes

gigxz commented on drivers/hmis/app/models/hmis/user.rb:
```diff
@@ -85,12 +90,12 @@ def lock_access!(opts = {})
   private def permissions_base_for_entity(entity)
-    return [entity.data_source, *entity.projects] if entity.is_a? Hmis::Hud::Client
+    return entity.projects if entity.is_a? Hmis::Hud::Client
     return entity if entity.is_a? Hmis::Hud::Organization
     return entity if entity.is_a? Hmis::Hud::Project
 
     if entity.is_a? Hmis::File
-      return [entity.client.data_source, *entity.client.projects] unless entity.enrollment.present?
+      return entity.client.data_source unless entity.enrollment.present?
```
I don't understand this change, can you explain?

If a user has access to see a client BECAUSE they can see a project that the client is enrolled in, then they should be able to see that client's files. I think this change takes that capability away. Is that right?

ZachCMP replied:
Yeah, I think this needs another look. I forgot that files are not connected to clients via enrollments. The intention was to base the permissions in the same way as client, so I'll go back and adjust to make that the case

gigxz replied:
Ok, then I think this should be `return entity.client.data_source unless entity.enrollment.present?`

gigxz commented on drivers/hmis/spec/requests/hmis/client_search_spec.rb:
```diff
@@ -44,16 +45,31 @@
+      expect(clients).to include({ 'id' => client3.id.to_s })
+    end
+
+    it 'should exclude clients enrolled at a project without user view permission' do
```
Can you add a test case for WIP enrollments as well? You can make an enrollment WIP by just calling `enrollment.save_in_progress`.

gigxz commented on drivers/hmis/app/graphql/schema.graphql:
```diff
@@ -404,12 +404,15 @@ type Client {
   canManageOwnClientFiles: Boolean!
   canViewAnyConfidentialClientFiles: Boolean!
   canViewAnyNonconfidentialClientFiles: Boolean!
+  canViewClient: Boolean!
```
Are there cases where GraphQL would resolve a Client that has `canViewClient: false`? 🤔

ZachCMP replied:
Probably not in our case. I'll remove it

gigxz committed:
Add unenrolled scope

gigxz commented on drivers/hmis/app/models/hmis/hud/client.rb:
```diff
@@ -54,9 +54,12 @@ class Hmis::Hud::Client < Hmis::Hud::Base
 
   scope :with_access, ->(user, *permissions, **kwargs) do
-    return none unless user.permissions?(*permissions, **kwargs)
+    base_scope = left_outer_joins(:projects).joins(:data_source)
 
-    joins(:data_source).merge(GrdaWarehouse::DataSource.hmis(user))
+    unenrolled_scope = user.permissions?(*permissions, **kwargs) ? base_scope.where(Project: { id: nil }).merge(GrdaWarehouse::DataSource.hmis(user)) : none
+    enrolled_scope = base_scope.merge(Hmis::Hud::Project.with_access(user, *permissions, **kwargs))
+
+    unenrolled_scope.or(enrolled_scope)
   end
```
This should use an `unenrolled` scope that considers a client "unenrolled" if they have NO enrollments (wip or otherwise): https://github.com/greenriver/hmis-warehouse/pull/3004/files#r1180348505

`enrolled_scope = base_scope.merge(..)` also needs to be updated to include projects that the client has WIP enrollments at.

gigxz committed:
Rm unused fn

eanders commented on drivers/hmis/app/models/hmis/hud/client.rb:
```diff
@@ -54,9 +54,12 @@ class Hmis::Hud::Client < Hmis::Hud::Base
+    base_scope = left_outer_joins(:projects).joins(:data_source)
 
-    joins(:data_source).merge(GrdaWarehouse::DataSource.hmis(user))
+    unenrolled_scope = user.permissions?(*permissions, **kwargs) ? base_scope.where(Project: { id: nil }).merge(GrdaWarehouse::DataSource.hmis(user)) : none
```
```suggestion
    unenrolled_scope = user.permissions?(*permissions, **kwargs) ? base_scope.where(p_t[:id].eq(nil)).merge(GrdaWarehouse::DataSource.hmis(user)) : none
```

eanders replied:
I'm not a fan of the active record nested where, especially when the table name differs from the standard rails snake case version.

gigxz replied:
It's not enough to look for nil project IDs anyway, because WIP Enrollments have nil project IDs. See proposed scope: https://github.com/greenriver/hmis-warehouse/pull/3004/files#r1180348505

ZachCMP replied:
@eanders Yeah, I meant to contact you about doing that the right way. I know nested is usually not the way to go in this project.

eanders commented on drivers/hmis/app/models/hmis/user.rb:
```diff
@@ -60,12 +60,17 @@ def load_effective_permissions
+      return true if base_entities.nil? && ::Hmis::Role.global_permissions.include?(permission)
+
+      raise "Invalid entity '#{entity.class.name}' for permission '#{permission}'" if base_entities.nil?
 
-      raise "Invalid entity '#{entity.class.name}' for permission '#{permission}'" unless base_entities.present?
+      return false unless base_entities.present?
```
Is it possible for base_entities to be empty but not nil?  I assume that'd be an empty array.  It seems odd that nil and [] would have different results, maybe a comment as to why would be appropriate?

ZachCMP replied:
Yes, those are two different cases.

If `base_entities` is `nil` it means that we couldn't find anything to base the permissions on. In the case of global permissions that's expected so we just check the global permission; In the case of non-global permissions that's an error indicating that we couldn't find an appropriate base for that entity.

If `base_entities` is `[]` it means that we found a valid base for the entity, but it doesn't have any of those base records. Therefore there it's not an error, but there are no records that would grant permissions on the entity, so it would return `false` to indicate that the user doesn't have the requested permissions on the entity.

I can add a comment to clarify

ZachCMP reviewed (commented):
@gigxz I didn't think we had to handle WIP enrollments based on the conversation in PT, which is why I didn't handle them here. It sounds like we actually do have to handle them, which is fine, I can go back and adjust it. Your PR with the `unenrolled` scope looks good to me and it'd be perfect for this case. Let me know when it's merged and I'll roll it in.

I probably won't get to updating this until Monday though, just FYI

gigxz commented:
@ZachCMP I opened https://github.com/greenriver/hmis-warehouse/pull/3004 against your branch so feel free to merge it in. The new scopes should help with most, if not all, of the WIP issues.

ZachCMP committed:
Merge branch 'release-73' into 185011051-closed-client-access

ZachCMP committed:
Merge remote-tracking branch 'origin/gig/185011051-closed-client-access' into 185011051-closed-client-access

ZachCMP committed:
Address PR feedback

ZachCMP committed:
Merge remote-tracking branch 'origin/release-73' into 185011051-closed-client-access

ZachCMP commented:
@gigxz Addressed the feedback above and pushed. Let me know what else is needed here. Also updated the frontend PR in kind, though that was only really schema changes

ZachCMP committed:
Merge branch 'release-73' into 185011051-closed-client-access

ZachCMP committed:
Merge branch '185011051-closed-client-access' of github.com:greenriver/hmis-warehouse into 185011051-closed-client-access

gigxz commented on drivers/hmis/app/models/hmis/user.rb:
```diff
@@ -84,16 +86,25 @@ def lock_access!(opts = {})
+    return entity if entity.is_a? Hmis::Hud::Organization
+    return entity if entity.is_a? Hmis::Hud::Project
+
+    return permissions_base_for_entity(entity.enrollment || entity.client) if entity.is_a? Hmis::File
```
nice

ZachCMP commented on drivers/hmis/app/models/hmis/user.rb:
```diff
@@ -84,16 +86,25 @@ def lock_access!(opts = {})
+      return entity.projects_including_wip if entity.enrolled?
 
-      return entity.enrollment.project
+      return Hmis::Hud::Project.viewable_by(self)
```
I also wanted to call out this change just to make sure you all agree with it. The logic here is that the permissions base for the client would be their projects if they are enrolled. If not enrolled, the base would be all projects the user can see, which would effectively mean that they have a permission on the client if they have that permission in any of the projects they can see. This is better than making it a special case in the `#{permission}_for?` method and achieves the same thing as far as I can tell.

I believe that works and will behave in the right way, but I wanted to call it out for a second opinion in case my thinking is incorrect here.

eanders replied:
I think that makes sense... I don't quite understand the returning of a user's projects, or how that would help determine if they can see the client that has no enrollments, but I agree, you can see the client if you have _some_ ability to see clients and the client has no enrollments.

gigxz replied:
Hmmm I think it makes sense, but the `Project.viewable_by` and subsequent access lookups seem unnecessary.  We don't need to do all that just to determine if you have that permission globally.

> This is better than making it a special case in the #{permission}_for? method and achieves the same thing as far as I can tell.

Is the alternative to return this here:
```
return entity.projects_including_wip if entity.is_a? Hmis::Hud::Client
```
and then on line 70 add:
```
# If Client has no enrollments, allow the permission if the user has it anywhere
# (check if its an array rather than if the entity is a client, because File may return client-based project list too..)
entity_is_unenrolled = base_entities.is_a?(ActiveRecord::Relation) && base_entities.empty? && (entity.is_a?(Client) || entity.is_a?(File))
return send("#{permission}?") if entity_is_unenrolled
```

?

If so, I think some version of the latter special case is probably preferred because we skip the unnecessary query to `Project.viewable_by(self)` which can be expensive. We need to build the project list, since we really just want to do a global check. Right?

ZachCMP replied:
Yeah, that's the rub for me. A global permission lookup is preferable, but I don't really want to add a special condition at line 70, as I think that makes that method harder to reason about. I think I'll add another method called `uses_global_permissions?(entity)` to determine whether the entity should use global permissions or not, then add the special cases to that. I think that's better. Do you agree?

gigxz replied:
Yeah, that works. As long as we are avoiding the extra db hit for a global project list. Thanks!

gigxz commented on drivers/hmis/app/models/hmis/hud/client.rb:
```diff
@@ -61,9 +64,14 @@ class Hmis::Hud::Client < Hmis::Hud::Base
 
   scope :with_access, ->(user, *permissions, **kwargs) do
-    return none unless user.permissions?(*permissions, **kwargs)
+    pids = Hmis::Hud::Project.with_access(user, *permissions, **kwargs).pluck(:id)
+    base_scope = left_outer_joins(:projects, :wip, :data_source)
+
+    unenrolled_scope = user.permissions?(*permissions, **kwargs) ? unenrolled.left_outer_joins(:data_source).merge(GrdaWarehouse::DataSource.hmis(user)) : none
+    enrolled_scope = base_scope.where(p_t[:id].in(pids))
+    wip_scope = base_scope.where(wip_t[:project_id].in(pids))
 
-    joins(:data_source).merge(GrdaWarehouse::DataSource.hmis(user))
+    unenrolled_scope.or(enrolled_scope).or(wip_scope)
   end
```
This seems to be working as intended, but I'm seeing a significant performance hit. My rails consoled timed out trying to evaluate this scope for a user that has access to a lot of projects.

I could think of two improvements, I see an impact but its pretty minor.
1) Get rid of some of the unnecessary Data source joins
2) Stop doing `left_outer_joins(:projects, :wip)` twice (once in this scope, once in "unenrolled" scope) - I have no idea if Rails would've been smart enough to collapse that already, I didn't look closely at the sql to compare

```
  scope :with_access, ->(user, *permissions, **kwargs) do
    pids = Hmis::Hud::Project.with_access(user, *permissions, **kwargs).pluck(:id)
    data_source_id = user.hmis_data_source_id
    base_scope = where(data_source_id: data_source_id).left_outer_joins(:projects, :wip)

    unenrolled_scope = user.permissions?(*permissions, **kwargs) ? base_scope.unenrolled : none
    enrolled_scope = base_scope.where(p_t[:id].in(pids))
    wip_scope = base_scope.where(wip_t[:project_id].in(pids))

    unenrolled_scope.or(enrolled_scope).or(wip_scope)
  end
```

@eanders do you have any other suggestion for speeding this up? I'm thinking we could cache something if needed, like project access list, but it's tricky since it's all based on permission.

gigxz commented on drivers/hmis/app/models/hmis/hud/client.rb:
```diff
@@ -61,9 +64,14 @@ class Hmis::Hud::Client < Hmis::Hud::Base
+    wip_scope = base_scope.where(wip_t[:project_id].in(pids))
 
-    joins(:data_source).merge(GrdaWarehouse::DataSource.hmis(user))
+    unenrolled_scope.or(enrolled_scope).or(wip_scope)
```
This is not returning a unique scope of clients, which I think it should. Do yall agree? Whats the most efficient way to to that? Elsewhere we pluck IDs and do another lookup.
@eanders @ZachCMP

ZachCMP replied:
We could try plucking the IDs from the three scopes and then doing `where(id: [*unenrolled_ids, *enrolled_ids, *wip_ids])`. Maybe that would help performance too?

eanders replied:
That was going to be my suggestion.

eanders replied:
or `where(id: unenrolled_ids + enrolled_ids + wip_ids)` which is easier to read for me.
`or` clauses can be slow.  If the plucking of the ids is significantly faster that seems like a good approach to me.

eanders replied:
Oh, I was confused, I was responding only to the performance question below, but yeah, maybe it'll solve both.

gigxz replied:
great

ZachCMP committed:
Fix PR feedback

ZachCMP committed:
Merge remote-tracking branch 'origin/release-73' into 185011051-closed-client-access

ZachCMP commented:
@gigxz I switched to plucking IDs and fixed the projects lookup. Take a look and let me know if this needs anything else.

gigxz committed:
Merge branch 'release-73' into 185011051-closed-client-access

gigxz committed:
Dont treat file as global if its attached to an enrollment

gigxz reviewed (approved):
thanks @ZachCMP!

gigxz commented:
I'm good with this, feel free to merge whenever you're ready @eanders
